p2p/local.go

```diff
@@ -0,0 +1,132 @@
+package p2p
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"gx/ipfs/QmNqRnejxJxjRroz7buhrjfU8i3yNBLa81hFtmf2pXEffN/go-multiaddr-net"
+	ma "gx/ipfs/QmUxSEGbv2nmYNnfXi7839wwQqTN3kwQeUxe8dTjZWZs7J/go-multiaddr"
+	"gx/ipfs/QmVf8hTAsLLFtn4WPCRNdnaF2Eag2qTBS6uR8AiHPZARXy/go-libp2p-peer"
+	tec "gx/ipfs/QmWHgLqrghM9zw77nF6gdvT9ExQ2RB9pLxkd8sDHZf1rWb/go-temp-err-catcher"
+	"gx/ipfs/QmXdgNhVEgjLxjUoMs5ViQL7pboAt3Y7V7eGHRiE4qrmTE/go-libp2p-net"
+	"gx/ipfs/QmZNkThpqfVXs9GNbexPrfBbXSLNYeKrE7jwFM2oqHbyqN/go-libp2p-protocol"
+)
+
+// localListener manet streams and proxies them to libp2p services
+type localListener struct {
+	ctx context.Context
+
+	p2p *P2P
+	id  peer.ID
+
+	proto protocol.ID
+	laddr ma.Multiaddr
+	peer  peer.ID
+
+	listener manet.Listener
+}
+
+// ForwardLocal creates new P2P stream to a remote listener
+func (p2p *P2P) ForwardLocal(ctx context.Context, peer peer.ID, proto protocol.ID, bindAddr ma.Multiaddr) (Listener, error) {
+	listener := &localListener{
+		ctx: ctx,
+
+		p2p: p2p,
+		id:  p2p.identity,
+
+		proto: proto,
+		laddr: bindAddr,
+		peer:  peer,
+	}
+
+	if err := p2p.Listeners.Register(listener); err != nil {
+		return nil, err
+	}
+
+	go listener.acceptConns()
+
+	return listener, nil
+}
+
+func (l *localListener) dial(ctx context.Context) (net.Stream, error) {
+	cctx, cancel := context.WithTimeout(ctx, time.Second*30) //TODO: configurable?
+	defer cancel()
+
+	return l.p2p.peerHost.NewStream(cctx, l.peer, l.proto)
+}
+
+func (l *localListener) acceptConns() {
+	for {
+		local, err := l.listener.Accept()
+		if err != nil {
+			if tec.ErrIsTemporary(err) {
+				continue
+			}
+			return
+		}
+
+		go l.setupStream(local)
+	}
+}
+
+func (l *localListener) setupStream(local manet.Conn) {
+	remote, err := l.dial(l.ctx)
+	if err != nil {
+		local.Close()
+		log.Warningf("failed to dial to remote %s/%s", l.peer.Pretty(), l.proto)
+		return
+	}
+
+	stream := &Stream{
+		Protocol: l.proto,
+
+		OriginAddr: local.RemoteMultiaddr(),
+		TargetAddr: l.TargetAddress(),
+
+		Local:  local,
+		Remote: remote,
+
+		Registry: l.p2p.Streams,
+	}
+
+	l.p2p.Streams.Register(stream)
+	stream.startStreaming()
+}
+
+func (l *localListener) start() error {
+	maListener, err := manet.Listen(l.laddr)
+	if err != nil {
+		return err
+	}
+
+	l.listener = maListener
+	return nil
+}
+
+func (l *localListener) Close() error {
```

Comment:
We may need a lock here. That depends on whether or not we close under a global lock.

Reply:
done

Reply:
Where?

Reply:
`l.p2p.Listeners.Deregister` now returns a boolean indicating whether the listener got deregistered (which happens with a lock, so this should only happen once). Would have been cleaner if I didn't do `git commit --amend --no-edit` all the time..

Reply:
Ah. So, the issue is that someone may concurrently try to read `listener`.

Reply:
It looks like there may still be a race. If there are two simultaneous close calls and a concurrent open call (for the same listener triplet) the one of the close calls may deregister the listener registered by the open call.

I think we need to pass the listener *into* `Deregister` and have deregister make sure it's the right one (compare pointers).

Reply:
This is a rather racy situation either way:

What can happen:
* close successfully; open; close freshly open
* close successfully; close fail; open;
* close successfully; open; close fail (with `listener didn't start yet` error)
* open fail; close successfully; close fail;
* Any other combination of those

I don't think other locking scheme would make this more 'defined' behavior